src/util.c

```diff
@@ -149,6 +149,26 @@ int write_zeroes(int fd, size_t l) {
         return 0; /* Return == 0 if we could only write out zeroes */
 }
 
+static bool memeqzero(const void *data, size_t length) {
+        /* Does the buffer consist entirely of NULs?
+         * Copied from https://github.com/rustyrussell/ccan/blob/master/ccan/mem/mem.c#L92.
+         */
+
+        const unsigned char *p = data;
+        size_t len;
+
+        /* Check first 16 bytes manually */
+        for (len = 0; len < 16; len++, length--) {
+                if (length == 0)
+                        return true;
+                if (*p++)
+                        return false;
+        }
+
+        /* Now we know first 16 bytes are NUL, memcmp with self.  */
+        return memcmp(data, p, length) == 0;
```

Comment:
urks, overlapping memcmp()… is the precise behaviour of that API? it appears to me that libcs are free to change their implementation of the implementing algorithm any time, so they could even compare from the end to the front if they like, or in fact any other order. And if they do that then all of this will just break apart horribly...

Moreover, 16 bytes in one go, does that really make sense? We live in a world where CPUs tend to implement AVX-512 these days, which means optimized string comparison routines could compare 64 bytes in one go, and any implementation that does that would not be compatible with the function above...

(This isn't a dream of the future even: example code to do string comparisons with AVX-512 exists already: https://github.com/WojciechMula/toys/tree/master/avx512-string it's only a matter of time until code like that enters glibc proper and the other libcs...)

Reply:
so, what about this: we can implement this without overlapping but still take benefit of memcmp()s speed, by comparing the first half of the memory area with the second one, and then recursively going down into its halves and so on. would be recursive of course, which sucks, but then again the depth is not going to be beyond 64...

we'd anchor the algorithm with comparing against a static const 64 byte array initialized to 0 and compiled in...

Reply:
The order of comparisons does not matter for correctness, just for efficiency (this just checks that the first 16 bytes are zero and are equals to the next 16 bytes, and so on). But I don't think glibc can realistically change the order here, it'd negatively affect too many things. And even if that ever happens, this code can be adjusted, since it's just an optimization. The only case where reverted comparison in memcmp would matter is when the buffer has repeated words, and those words are non-zero, which shouldn't happen that often.

> AVX-512

Sure, go for it ;) My implementation has the advantage of being simple, while still being good enough that `loop_write_with_holes` stops showing up in the first stop in `perf report`. Optimizing it more at this point would be premature, imho, since it's other parts that are slower now.

> so, what about this: we can implement this without overlapping but still take benefit of memcmp()s speed, by comparing the first half of the memory area with the second one, and then recursively going down into its halves  and so on

That would trash the cache, and require many more library calls, _both_ of which are likely to kill efficiency.

Reply:
hmm, if we go for such a bisection algorithm i figure  any non-64 aligned prefixed should be compared separately though...

Reply:
> The order of comparisons does not matter for correctness, just for efficiency (this just checks that the first 16 bytes are zero and are equals to the next 16 bytes, and so on). 

Ah, indeed, I guess that means it is OK then. But maybe add a comment about the expectations here...

> But I don't think glibc can realistically change the order here, it'd negatively affect too many things. 

Well, there was that overlapping memcmp() kerfuffle a few years ago, and to my knowledge they left the change in for good, no?

Reply:
> Well, there was that overlapping memcmp() kerfuffle a few years ago, and to my knowledge they left the change in for good, no?

Right. I forgot about that, and I can't find the reference. I can't find the discussion though, do you have some link?